vyper/venom/passes/common_subexpression_elimination.py

```diff
@@ -0,0 +1,81 @@
+from vyper.utils import OrderedSet
+from vyper.venom.analysis.available_expression import (
+    _UNINTERESTING_OPCODES,
+    AvailableExpressionAnalysis,
+)
+from vyper.venom.analysis.dfg import DFGAnalysis
+from vyper.venom.analysis.liveness import LivenessAnalysis
+from vyper.venom.basicblock import IRBasicBlock, IRInstruction, IRVariable
+from vyper.venom.passes.base_pass import IRPass
+
+_MAX_DEPTH = 5
+_MIN_DEPTH = 2
+
+
+class CSE(IRPass):
+    available_expression_analysis: AvailableExpressionAnalysis
+
+    def run_pass(self, min_depth: int = _MIN_DEPTH, max_depth: int = _MAX_DEPTH):
+        available_expression_analysis = self.analyses_cache.request_analysis(
+            AvailableExpressionAnalysis, min_depth, max_depth
+        )
+        assert isinstance(available_expression_analysis, AvailableExpressionAnalysis)
+        self.available_expression_analysis = available_expression_analysis
+
+        while True:
+            replace_dict = self._find_replaceble()
+            if len(replace_dict) == 0:
+                return
+            self._replace(replace_dict)
+            self.analyses_cache.invalidate_analysis(DFGAnalysis)
+            self.analyses_cache.invalidate_analysis(LivenessAnalysis)
+            available_expression_analysis = self.analyses_cache.force_analysis(
+                AvailableExpressionAnalysis, min_depth, max_depth
+            )
+            assert isinstance(available_expression_analysis, AvailableExpressionAnalysis)
+            self.available_expression_analysis = available_expression_analysis
+
+    # return instruction and to which instruction it could
+    # replaced by
+    def _find_replaceble(self) -> dict[IRInstruction, IRInstruction]:
+        res: dict[IRInstruction, IRInstruction] = dict()
+        for bb in self.function.get_basic_blocks():
+            for inst in bb.instructions:
+                if inst in _UNINTERESTING_OPCODES:
+                    continue
+                inst_expr = self.available_expression_analysis.get_expression(inst)
+                avail = self.available_expression_analysis.get_available(inst)
```

Comment:
How about `get_available()` doing directly returning the `first_inst` if there is one or `None` and avoid the need to use `get_expression()` before?